storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerAppendInfo.java

```
@@ -119,6 +119,21 @@ private void checkProducerEpoch(short producerEpoch, long offset, short transact
         boolean invalidEpoch = (transactionVersion >= 2) ? (producerEpoch <= current) : (producerEpoch < current);
 
         if (invalidEpoch) {
+            // TV2 Idempotent Retry Detection:
+            // When markerEpoch == currentEpoch and no transaction is ongoing, this is a retry
+            // of a marker that was already successfully written. Common scenarios:
+            // 1. Coordinator recovery: reloading PREPARE_COMMIT/ABORT from the transaction log
+            // 2. Network retry: marker was written but response was lost due to disconnection
+            // In both cases, the transaction has already ended (currentTxnFirstOffset is empty),
+            // so we can safely treat this as idempotent success.
+            if (transactionVersion >= 2 &&
+                    producerEpoch == current &&
+                    updatedEntry.currentTxnFirstOffset().isEmpty()) {
+                log.debug("Received duplicate transaction marker for producer {} with epoch {} " +
+                                "but transaction is no longer ongoing, treating as idempotent success",
+                        producerId, producerEpoch);
+                throw new IdempotentTransactionMarkerException();
```

Hmm -- I'm not sure how I feel about throwing this error here. Why did we choose to throw an error?

---

Ideally we would try to avoid an error thrown and control flow checking errors in a case where there is no error.

---

I agree that using exception for control is an anti-pattern, especially for success scenarios.

However, the write path for the offset topic appears to lack built-in idempotency logic, so avoiding the exception approach might require more refactoring. Regardless, I agree it is worth the effort to implement this without relying on exceptions

---

If we simply return early from `ProducerAppendInfo#checkProducerEpoch`, we still need to prevent duplicate data from being written to the log. The validation happens before the actual append operation, so just returning allows the code to continue and write the duplicate marker anyway.

The general produce path can handle duplicate writes more easily through conditional logic after validation. However, the write path for the offset topic appears to lack built-in idempotency logic. Therefore, throwing an exception provides a clean way to immediately break the append flow once we detect an idempotent retry. The exception propagates up through the call stack, preventing the duplicate write regardless of which code path is handling the marker. The caller can then catch this specific exception type and treat it as a successful operation.

While using exceptions for control flow is not ideal, it's a pragmatic solution given the current code structure.

---

To avoid using exceptions for control flow, an alternative approach is to simply suppress the `InvalidProducerEpochException` when the "safe" conditions are met (i.e., the transaction is completed, the protocol is v2, and the producer epochs are identical)


While this results in appending a duplicate marker to the log, it should be acceptable as this scenario is rare and the record size is negligible. Furthermore, both the broker and clients can handle such duplicate markers gracefully

@m1a2st @jolshan WDYT?

---

I think it is ok to write a duplicate marker as long as we have the log lock and can't write anything else (ie no transaction will start)

---

> While this results in appending a duplicate marker to the log, it should be acceptable as this scenario is rare and the record size is negligible. Furthermore, both the broker and clients can handle such duplicate markers gracefully

Agreed. Suppressing the exception and allowing the duplicate marker is simpler and more pragmatic. The duplicate marker is harmless and this scenario is rare enough that the tradeoff is worth it for cleaner code.

---

Awesome, the code is also looking a lot cleaner. 👍